src/main/java/co/cask/gcp/datastore/sink/DatastoreSinkTransformer.java

```diff
@@ -0,0 +1,326 @@
+package co.cask.gcp.datastore.sink;
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.gcp.datastore.sink.util.SinkKeyType;
+import co.cask.gcp.datastore.util.DatastorePropertyUtil;
+import com.google.cloud.Timestamp;
+import com.google.cloud.datastore.Blob;
+import com.google.cloud.datastore.BlobValue;
+import com.google.cloud.datastore.BooleanValue;
+import com.google.cloud.datastore.DoubleValue;
+import com.google.cloud.datastore.EntityValue;
+import com.google.cloud.datastore.FullEntity;
+import com.google.cloud.datastore.IncompleteKey;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.LongValue;
+import com.google.cloud.datastore.NullValue;
+import com.google.cloud.datastore.PathElement;
+import com.google.cloud.datastore.StringValue;
+import com.google.cloud.datastore.TimestampValue;
+import com.google.cloud.datastore.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Transforms {@link StructuredRecord} to the Google Cloud Datastore {@link FullEntity}
+ * using {@link DatastoreSinkConfig}.
+ */
+public class DatastoreSinkTransformer {
+  private static final Logger LOG = LoggerFactory.getLogger(DatastoreSinkTransformer.class);
+
+  private final DatastoreSinkConfig config;
+
+  public DatastoreSinkTransformer(DatastoreSinkConfig config) {
+    this.config = config;
+  }
+
+  public FullEntity<?> transformStructuredRecord(StructuredRecord record) {
+    KeyFactory keyFactory = new KeyFactory(config.getProject())
+      .setNamespace(config.getNamespace())
+      .setKind(config.getKind());
+
+    String keyAlias = config.getKeyAlias();
+    FullEntity.Builder<IncompleteKey> entityBuilder = FullEntity.newBuilder();
+    List<Schema.Field> fields = Objects.requireNonNull(record.getSchema().getFields(), "Schema fields cannot be empty");
+
+    boolean useAutoGeneratedKey = config.shouldUseAutoGeneratedKey();
+    for (Schema.Field field : fields) {
+      String fieldName = field.getName();
+      if (!useAutoGeneratedKey && keyAlias.equals(fieldName)) {
+        entityBuilder.setKey(convertToKey(record, keyFactory, field));
+      } else {
+        boolean excludeFromIndex = isExcludedFromIndex(fieldName);
+        entityBuilder.set(fieldName, convertToValue(fieldName, field.getSchema(), record, excludeFromIndex));
+      }
+    }
+
+    if (useAutoGeneratedKey) {
+      entityBuilder.setKey(getAutoGeneratedKey(keyFactory));
+    }
+    return entityBuilder.build();
+  }
+
+  private boolean isExcludedFromIndex(String fieldName) {
+    switch (config.getIndexStrategy()) {
+      case ALL:
+        return false;
+      case NONE:
+        return true;
+      case CUSTOM:
+        return !config.getIndexedProperties().contains(fieldName);
+      default:
+        throw new IllegalStateException(
```

Comment:
this can be made a little cleaner and more efficient by setting a Predicate in the constructor instead of switching on the index strategy for every field of every record. Something like:

```
Predicate<String> isExcludedFromIndex;
...

switch (config.getIndexStrategy()) {
  case ALL:
    isExcludedFromIndex = fieldName -> false;
    break;
  case NONE:
    isExcludedFromIndex = fieldName -> true;
    break;
  case CUSTOM:
    isExcludedFromIndex = fieldName -> !config.getIndexedProperties().contains(fieldName);
    break;
  default:
    throw ...
}
```